File: Day 5/Task 1/Task 1.cpp
#include <iostream>
#include <fstream>
#include <stack>
#include <vector>
#include <string>
#include <iterator>
#include <sstream>
#include <regex>


std::vector<std::stack<char>> ReadCrates(const std::vector<std::string>& fileData) {
	auto it = fileData.rbegin();
	
	std::stringstream buffer;
	buffer << *it++;
	int maxStackIndex;

	while (buffer >> maxStackIndex);

	std::vector<std::stack<char>> stackOfCrates(maxStackIndex);
	while (it != fileData.rend()) {

		for (int i = 0; i < maxStackIndex; i++) {
			if (std::isalpha(it->at(4 * i + 1)))
				stackOfCrates[i].push(it->at(4 * i + 1));
		}
		++it;
	}
	return stackOfCrates;

}


int main()
{
	std::ifstream fhandle("..\\day_5_1.txt");
	std::vector<std::string> fileData;
	std::string line;

	while (std::getline(fhandle, line), line != "")
		fileData.push_back(line);
	
	std::vector<std::stack<char>> stackOfCrates = ReadCrates(fileData);
	
	std::string moveInstruction;
	std::regex pattern("move ([0-9]+) from ([0-9]+) to ([0-9]+)");
	std::smatch instructionParams;

	while (std::getline(fhandle, moveInstruction)) {

		if (std::regex_search(moveInstruction, instructionParams, pattern)) {

			int numOfMoves = std::stoi(instructionParams[1]);
			int stackToPop = std::stoi(instructionParams[2]) - 1;
			int stackToPush = std::stoi(instructionParams[3]) - 1;

			for (int i = 0; i < numOfMoves; i++) {
				stackOfCrates[stackToPush].push(stackOfCrates[stackToPop].top());
				stackOfCrates[stackToPop].pop();
			}
		}
	}

	std::cout << "The crates that are on top of each stack are: ";
	for (auto it = stackOfCrates.begin(); it != stackOfCrates.end(); it++)
		std::cout << "[" << it->top() << "]" << " ";

}

File: Day 5/Task 1/Task_1.py
from Stack import Stack
import re

def read_crates(file_data):
    max_stack_index = int(file_data[-1].split()[-1])
    stack_of_crates = [Stack([]) for i in range(max_stack_index)]

    for i, line in enumerate(reversed(file_data)):
        if i == 0:
            continue

        for j in range(max_stack_index):

            if line[4*j+1].isalpha():
                stack_of_crates[j].push(line[4*j+1])

    return stack_of_crates

file_data = list()
with open(r"..\day_5_1.txt") as fhandle:
    for line in fhandle:
        #get rid of the \n only
        line = line[:-1]

        if line == '':
            break
        file_data.append(line)
    stack_of_crates = read_crates(file_data)

    regex_pattern = "move ([0-9]+) from ([0-9]+) to ([0-9]+)"
    for line in fhandle:
        line = line[:-1]
        instruction_params = re.search(regex_pattern, line)
        num_of_moves, stack_to_pop, stack_to_push = tuple(int(param) for param in instruction_params.groups())

        for i in range(num_of_moves):
            stack_of_crates[stack_to_push-1].push(stack_of_crates[stack_to_pop-1].top())
            stack_of_crates[stack_to_pop-1].pop()

    print("The crates that are on top of each stack are: ")
    for stack in stack_of_crates:
        print(f"[{stack.top()}] ", end = '')

